crates/rollup/src/driver/context/mod.rs

```diff
@@ -0,0 +1,148 @@
+            ChainNotification::Revert { .. } => None,
+        }
+    }
+
+    pub fn reverted_chain(&self) -> Option<Blocks> {
+        match self {
+            ChainNotification::Revert { old_blocks } => Some(old_blocks.clone()),
+            ChainNotification::Reorg { old_blocks, .. } => Some(old_blocks.clone()),
+            ChainNotification::New { .. } => None,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct Blocks(Arc<BTreeMap<BlockNumber, Block<TxEnvelope>>>);
+
+impl Blocks {
+    pub fn tip(&self) -> BlockNumber {
+        *self.0.last_key_value().expect("Blocks should have at least one block").0
+    }
+
+    pub fn fork_block(&self) -> BlockNumber {
+        let first = self.0.first_key_value().expect("Blocks should have at least one block").0;
+        first.saturating_sub(1)
+    }
+}
+
+impl From<Block<TxEnvelope>> for Blocks {
+    fn from(value: Block<TxEnvelope>) -> Self {
+        let mut blocks = BTreeMap::new();
+        blocks.insert(value.header.number.unwrap(), value);
+        Self(Arc::new(blocks))
+    }
+}
+
+impl From<Vec<Block<TxEnvelope>>> for Blocks {
+    fn from(value: Vec<Block<TxEnvelope>>) -> Self {
+        let mut blocks = BTreeMap::new();
+        for block in value {
+            blocks.insert(block.header.number.unwrap(), block);
+        }
+        Self(Arc::new(blocks))
+    }
+}
+
+impl From<Arc<Chain>> for Blocks {
+    fn from(value: Arc<Chain>) -> Self {
+        // annoyingly, we must convert from reth::primitives::SealedBlockWithSenders
+        // to alloy::rpc::types::Block
+
+        let mut blocks = BTreeMap::new();
+        for (block_number, sealed_block) in value.blocks() {
+            let block = Block {
+                header: parse_reth_rpc_header(sealed_block),
+                uncles: sealed_block.ommers.iter().map(|x| x.hash_slow()).collect(),
+                transactions: BlockTransactions::Full(
+                    sealed_block.transactions().flat_map(reth_to_alloy_tx).collect(),
+                ),
+                size: Some(U256::from(sealed_block.size())),
+                withdrawals: None, // TODO: parse withdrawals
+                other: OtherFields::default(),
+            };
+            blocks.insert(*block_number, block);
+        }
+        Self(Arc::new(blocks))
+    }
+}
+
+impl From<ExExNotification> for ChainNotification {
+    fn from(value: ExExNotification) -> Self {
+        match value {
+            ExExNotification::ChainCommitted { new } => Self::New { new_blocks: new.into() },
+            ExExNotification::ChainReverted { old } => Self::Revert { old_blocks: old.into() },
+            ExExNotification::ChainReorged { old, new } => {
+                Self::Reorg { old_blocks: old.into(), new_blocks: new.into() }
+            }
+        }
+    }
+}
+
+// from reth::primitives::SealedBlock to alloy::rpc::types::Header
+fn parse_reth_rpc_header(block: &reth::primitives::SealedBlock) -> Header {
```

Comment:
Let's write up tickets to upstream these conversion utilities to reth. Reth should provide a `From` impl into `alloy::rpc::types::Header`.

Reply:
Cool, I'll collect all of them and prepare a ticket!